### DISCUSSION TITLE
How to use bevy_reflect to get component field values at runtime?

### DISCUSSION THREAD

**Discussion by DumplingEater** [bevyengine/bevy#7886] [Q&A] [answered]
I've spent some time looking around the docs, various issues, and reading through the code and can't quite figure out how to do this. For context, I am just messing around building a UI / editor, and am trying to create functionality where you can click on an entity, it shows the entities components, then you can click a component, and it expands to show all reflected fields and their values with the goal of editing those values.

As a somewhat distilled example, here is what I have so far. I have another component that builds the ui for a scene-graph, and when you click an entity it sets that Entity type on the ui_state resource I have. This is the system that would draw that selected entity's field info on another panel.

```rust
#[derive(Debug, Copy, Clone, Serialize, Deserialize, Reflect, FromReflect, Resource)]
#[repr(C)]
pub struct Vector3f{
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

#[derive(Component, Debug, Clone, Serialize, Deserialize, Reflect, FromReflect)]
#[reflect(Component)]
pub struct TransformComponent{
    pub global_position: Vector3f,
}

// some time later in the code
let registry_arc = &world.get_resource_mut::<TypeRegistryResource>().unwrap().0;
let mut registry = registry_arc.write();
registry.register::<TransformComponent>();
registry.register::<Vector3f>();

// and here is the system
pub fn EntityInspectionUiSystem(
    query: Query<(&EntityInspectorComponent, Entity)>,
    ui_state: Res<EditorUiState>,
    type_registry_res: Res<TypeRegistryResource>,
    world: &World,
){
    if let Some(selected_entity: Entity) = ui_state.selected_entity{
        // boilerplate to get the proper panel ui here
       let mut ui = some_boiler_plate();

        ui.heading(format!("Selected Entity {}", selected_entity.to_bits()));
        ui.separator();

        let archetypes = world.archetypes();
        let components = world.components();
        for archetype in archetypes.iter() {
            if archetype.entities().iter().any(|e| e.entity() == selected_entity) {
                for comp_id in archetype.components() {
                    if let Some(comp_info) = components.get_info(comp_id) {
                        let component_name = comp_info.name().split("::").last().unwrap();
                        // at this point we have the component name
                        egui::CollapsingHeader::new(component_name).show(&mut ui, |ui|{
                            // I am not convinced I am actually doing this the best way, but it's what I managed to figure out
                            let registration = type_registry
                                .get_with_name(comp_info.name())
                                .unwrap();
                            let reflect_component = registration.data::<ReflectComponent>().unwrap();
                            let dyn_reflect = reflect_component.reflect(world, selected_entity).unwrap();
                            let type_info = dyn_reflect.get_type_info();
                                match type_info {
                                    TypeInfo::Struct(struct_info) => {
                                        for (i, field) in struct_info.iter().enumerate(){
                                            ui.horizontal(|ui|{
                                                ui.label(field.name());
                                            });
                                        }
                                    },
                                    _ => ();
                                }
                        });
                    }
                }
            }
        }
    }
}
```

So things I have considered are building a proxy dynamic struct and adding it to the entity this ui lives on for the duration of time that the component is being edited, reading field names off of the type info, checking if they're equal to certain types using NamedField.is::<f32>() etc, then, since I'd know the type at that point, assigning default values to the proxy DynamicStruct, having the ui operate on the proxy struct since you can pull field values off it, then applying that proxy struct to the reflected component on the selected entity.

Also if I could somehow cast the &dyn Reflect value or the ReflectComponent into a dynamic struct then I could probably just edit it directly, but couldn't figure out how to do that.

That seems more or less fine, but I wasn't super content with the idea of manually checking if a field was any possible type that my fields could be both when I'm building the proxy struct and when I'm reading fields off of it to build specific ui editors, which was what led me to ask this question. If I am going about this completely wrong, I am open to any and all advice.

To summarize, I basically am trying to figure out how to edit the fields of any given component on any given entity via a system without knowing the types of the components or fields at compile time.

**Comment by DumplingEater** [comment, marked as answer]
Just an update for anyone else who was curious, I was able to solve this by using `ReflectRef` to convert my `dyn Reflect` object into the a trait object of the subtrait `Struct`, and access the fields that way. Turns out it was literally on the landing page for bevy_reflect docs. Anyways here's the snippet I used:

```rust
let reflected = reflect_component.reflect(world, selected_entity).unwrap();
let ReflectRef::Struct(reflected) = reflected.reflect_ref() else { unreachable!() };
for (i, field) in reflected.iter_fields().enumerate() {
    ui.label(format!("{:?} : {:?}", reflected.name_at(i).unwrap(), reflected.field_at(i).unwrap()));
}
```

Obviously there's some mutability changes to be made to actually edit the fields, but hopefully this helps anyone trying to do something similar.

**Comment by wbpaley** [comment]
Hi @DumplingEater!  Still remember this stuff a year+ later?

I want the same thing: to iterate through my entities and get the values in their components. Your post(s) were _very_ helpful--as close as I've gotten. 

Unfortunately I get the field values for only two of the components--and can't see how to get the values for the others--can't see why I hit the `continue;` for some and not others. 

Any ideas?

Here's my munge of your code, followed by a grab from the console showing the kind of things I get: always get TypeInfo, but can't always get the ReflectRef for the values.

```
pub fn entity_inspection_system(
    mut commands: Commands,
    world: &World,
    mut examine_components_q: Query<(&ExamineComponents)>,
) {
    for (ec) in examine_components_q.iter_mut() {
        commands.entity(ec.entity).remove::<ExamineComponents>();

        let type_registry = world.get_resource::<AppTypeRegistry>().unwrap().clone();
        let type_registry = type_registry.deref().read();
        // for i in type_registry.iter() { println!("{:?}", i); }

        let archetypes = world.archetypes();
        let components = world.components();
        println!("\n\n\n\n");
        for archetype in archetypes.iter() {
            if archetype.entities().iter().any(|e| e.id() == ec.entity) {
                for comp_id in archetype.components() {
                    if let Some(comp_info) = components.get_info(comp_id) {
                        let type_key = comp_info.name();
                        println!("\n{:?}", type_key);
                        let registration = type_registry
                            .get_with_type_path(type_key);  // was: .get_with_name(comp_info.name())

                        if let Some(registration2) = registration {
                            let reflect_component = registration2.data::<ReflectComponent>().unwrap();
                            let original_entity = world.entity(ec.entity);
                            let reflected = reflect_component.reflect(original_entity).unwrap();

                            println!("type info: {:?}\n", reflected.deref().get_represented_type_info());

                            let ReflectRef::Struct(reflect_ref) = reflected.reflect_ref() else {
                                continue;  // was: unreachable!()
                            };
                            for (i, field) in reflect_ref.iter_fields().enumerate() {
                                println!("    {:?} : {:?}", reflect_ref.name_at(i).unwrap(), reflect_ref.field_at(i).unwrap());
                            }
                        }
                    }
                }
            }
        }
    }
}
```

And I get:

> "bevy_asset::handle::Handle<bevy_render::mesh::mesh::Mesh>"
> type info: Some(Enum(EnumInfo { type_path: TypePathVtable { type_path: "bevy_asset::handle::Handle<bevy_render::mesh::mesh::Mesh>", short_type_path: "Handle<Mesh>", type_ident: Some("Handle"), crate_name: Some("bevy_asset"), module_path: Some("bevy_asset::handle") }, type_id: TypeId { t: (9366316892559051162, 2539145655592770637) }, variants: [Tuple(TupleVariantInfo { name: "Strong", fields: [UnnamedField { index: 0, type_path: TypePathVtable { type_path: "std::sync::Arc<bevy_asset::handle::StrongHandle>", short_type_path: "Arc<StrongHandle>", type_ident: Some("Arc"), crate_name: Some("std"), module_path: Some("std::sync") }, type_id: TypeId { t: (12167382370610639244, 17279274000469605759) }, custom_attributes: {} }], custom_attributes: {} }), Tuple(TupleVariantInfo { name: "Weak", fields: [UnnamedField { index: 0, type_path: TypePathVtable { type_path: "bevy_asset::id::AssetId<bevy_render::mesh::mesh::Mesh>", short_type_path: "AssetId<Mesh>", type_ident: Some("AssetId"), crate_name: Some("bevy_asset"), module_path: Some("bevy_asset::id") }, type_id: TypeId { t: (5073353685420291322, 11399611058025075035) }, custom_attributes: {} }], custom_attributes: {} })], variant_names: ["Strong", "Weak"], variant_indices: {"Strong": 0, "Weak": 1}, custom_attributes: {} }))
> 
> 
> "bevy_asset::handle::Handle<bevy_pbr::pbr_material::StandardMaterial>"
> type info: Some(Enum(EnumInfo { type_path: TypePathVtable { type_path: "bevy_asset::handle::Handle<bevy_pbr::pbr_material::StandardMaterial>", short_type_path: "Handle<StandardMaterial>", type_ident: Some("Handle"), crate_name: Some("bevy_asset"), module_path: Some("bevy_asset::handle") }, type_id: TypeId { t: (2190834502334798386, 5526016345561453841) }, variants: [Tuple(TupleVariantInfo { name: "Strong", fields: [UnnamedField { index: 0, type_path: TypePathVtable { type_path: "std::sync::Arc<bevy_asset::handle::StrongHandle>", short_type_path: "Arc<StrongHandle>", type_ident: Some("Arc"), crate_name: Some("std"), module_path: Some("std::sync") }, type_id: TypeId { t: (12167382370610639244, 17279274000469605759) }, custom_attributes: {} }], custom_attributes: {} }), Tuple(TupleVariantInfo { name: "Weak", fields: [UnnamedField { index: 0, type_path: TypePathVtable { type_path: "bevy_asset::id::AssetId<bevy_pbr::pbr_material::StandardMaterial>", short_type_path: "AssetId<StandardMaterial>", type_ident: Some("AssetId"), crate_name: Some("bevy_asset"), module_path: Some("bevy_asset::id") }, type_id: TypeId { t: (18225880721845292907, 1867417461041737469) }, custom_attributes: {} }], custom_attributes: {} })], variant_names: ["Strong", "Weak"], variant_indices: {"Strong": 0, "Weak": 1}, custom_attributes: {} }))
> 
> 
> "bevy_transform::components::transform::Transform"
> type info: Some(Struct(StructInfo { type_path: TypePathVtable { type_path: "bevy_transform::components::transform::Transform", short_type_path: "Transform", type_ident: Some("Transform"), crate_name: Some("bevy_transform"), module_path: Some("bevy_transform::components::transform") }, type_id: TypeId { t: (1460694120476957121, 1930090812988915733) }, fields: [NamedField { name: "translation", type_path: TypePathVtable { type_path: "glam::Vec3", short_type_path: "Vec3", type_ident: Some("Vec3"), crate_name: Some("glam"), module_path: Some("glam") }, type_id: TypeId { t: (1912926410106069635, 2963730798302158443) }, custom_attributes: {} }, NamedField { name: "rotation", type_path: TypePathVtable { type_path: "glam::Quat", short_type_path: "Quat", type_ident: Some("Quat"), crate_name: Some("glam"), module_path: Some("glam") }, type_id: TypeId { t: (16042903574863222822, 4997740059470737957) }, custom_attributes: {} }, NamedField { name: "scale", type_path: TypePathVtable { type_path: "glam::Vec3", short_type_path: "Vec3", type_ident: Some("Vec3"), crate_name: Some("glam"), module_path: Some("glam") }, type_id: TypeId { t: (1912926410106069635, 2963730798302158443) }, custom_attributes: {} }], field_names: ["translation", "rotation", "scale"], field_indices: {"rotation": 1, "scale": 2, "translation": 0}, custom_attributes: {} }))
> 
>     "translation" : Vec3(0.0, 0.0, 16.5)
>     "rotation" : Quat(0.0, 1.0, -0.0, -0.0)
>     "scale" : Vec3(1.0, 1.0, 1.0)
> 
> "bevy_transform::components::global_transform::GlobalTransform"
> type info: Some(TupleStruct(TupleStructInfo { type_path: TypePathVtable { type_path: "bevy_transform::components::global_transform::GlobalTransform", short_type_path: "GlobalTransform", type_ident: Some("GlobalTransform"), crate_name: Some("bevy_transform"), module_path: Some("bevy_transform::components::global_transform") }, type_id: TypeId { t: (5537590884290836908, 5807618773129278876) }, fields: [UnnamedField { index: 0, type_path: TypePathVtable { type_path: "glam::Affine3A", short_type_path: "Affine3A", type_ident: Some("Affine3A"), crate_name: Some("glam"), module_path: Some("glam") }, type_id: TypeId { t: (15729112105241451245, 2619321799313709626) }, custom_attributes: {} }], custom_attributes: {} }))
> 

(and more...)
